src/maingoblint.ml

```diff
@@ -209,11 +211,16 @@ let basic_preprocess ~all_cppflags fname =
   in
   Preprocessor.FpathH.replace basic_preprocess_counts basename (count + 1);
   let nname = Fpath.append (GoblintDir.preprocessed ()) (Fpath.add_ext ".i" unique_name) in
-  (* Preprocess using cpp. *)
-  let arguments = all_cppflags @ Fpath.to_string fname :: "-o" :: Fpath.to_string nname :: [] in
-  let command = Filename.quote_command (Preprocessor.get_cpp ()) arguments in
-  if get_bool "dbg.verbose" then print_endline command;
-  (nname, Some {ProcessPool.command; cwd = None})
+  if ext <> ".i" && not (GobConfig.get_bool "pre.skipcpp") then
+    (* Preprocess using cpp. *)
+    let arguments = all_cppflags @ Fpath.to_string fname :: "-o" :: Fpath.to_string nname :: [] in
+    let command = Filename.quote_command (Preprocessor.get_cpp ()) arguments in
+    if get_bool "dbg.verbose" then print_endline command;
+    (nname, Some {ProcessPool.command; cwd = None})
+  else
+    (* No preprocessing needed. *)
+    (FileUtil.cp [(Fpath.to_string fname)] (Fpath.to_string nname);
+     (nname, None))
```

Comment:
The copying is completely unnecessary here though, we could just use the original file directly. The copy goes to the temporary `.goblint` directory, which is deleted after preprocessing anyway.

So we could avoid dragging in another dependency.

Reply:
We do have the fileutils library from #993 now, but even then this copying can be avoided.

Reply:
I thought keeping it this way would make it more consistent with `pre.keep`: Otherwise some files would be in the corresponding directory, and some would remain at their original path.
I don't think copying files is too expensive anyway.

Reply:
The directory is `.goblint/preprocessed` though, which would no longer be precisely true.

The bigger problem with copying is that the file paths in locations become incorrect: if the file isn't actually preprocessed from its original path into that directory, but is copied verbatim, then there are no `#line`, etc pragmas to indicate the original paths for CIL. Instead, it will be using the file's "actual" path, i.e. `.goblint/preprocessed/something.c`, which isn't where the `something.c` file originally is for the user. So the paths will be wrong in messages and not correctly displayable in GobPie.

Reply:
True. I have now removed the copying in this case.